restapi/auth.go

```diff
@@ -11,38 +13,85 @@ import (
-	"golang.org/x/crypto/bcrypt"
-	"gorm.io/gorm"
 )
 
+func generateStateOauthCookie(w http.ResponseWriter) (string, error) {
+	duration, err := config.GetTokenExpiryDuration()
+	if err != nil {
+		return "", err
+	}
+	var expiration = time.Now().Add(duration)
+	secure := (config.GetMode() == config.MODE_PROD)
+
+	b := make([]byte, 16)
+	rand.Read(b)
+	state := base64.URLEncoding.EncodeToString(b)
+	cookie := http.Cookie{Name: "oauthstate", Value: state, Expires: expiration, HttpOnly: true, Secure: secure}
+	http.SetCookie(w, &cookie)
+
+	return state, nil
+}
+
 func (api api) Login(w http.ResponseWriter, r *http.Request) render.Renderer {
-	var user models.User
-	err := json.NewDecoder(r.Body).Decode(&user)
+	state, err := generateStateOauthCookie(w)
+	if err != nil {
+		return rest.ErrInternal(api.logger, err)
+	}
+
+	url := api.oauthconfig.AuthCodeURL(state)
+	http.Redirect(w, r, url, http.StatusTemporaryRedirect)
+
+	return rest.JSONStatusOK(url)
+}
+
+func (api api) AuthCallback(w http.ResponseWriter, r *http.Request) render.Renderer {
+	expectedState, _ := r.Cookie("oauthstate")
+	if expectedState.Value != r.FormValue("state") {
+		return rest.ErrUnauthorized("Invalid oauth state")
+	}
+
+	// exchange received authorization code for an access token
+	token, err := api.oauthconfig.Exchange(context.Background(), r.FormValue("code"))
 	if err != nil {
-		rest.ErrInvalidRequest(api.logger, "Invalid login payload", err)
+		return rest.ErrUnauthorized("Invalid authorization code")
 	}
 
-	// Find user with username
-	var dbUser models.User
-	if err = api.database.Where("username=?", user.Username).First(&dbUser).Error; err != nil {
-		if err == gorm.ErrRecordNotFound {
-			return rest.ErrNotFound(fmt.Sprintf("Could not find user with username %s", user.Username))
-		}
+	// use access token to create a client which we use to access user info
+	client := api.oauthconfig.Client(context.Background(), token)
+	response, err := client.Get("https://www.googleapis.com/oauth2/v3/userinfo")
+	if err != nil {
+		return rest.JSONStatusOK("Invalid access token")
+	}
+	defer response.Body.Close()
+	user := models.User{}
+	err = json.NewDecoder(response.Body).Decode(&user)
+	if err != nil {
 		return rest.ErrInternal(api.logger, err)
 	}
 
-	// Validate password
-	if err = bcrypt.CompareHashAndPassword([]byte(dbUser.Password), []byte(user.Password)); err != nil {
-		return rest.ErrUnauthorized(fmt.Sprintf("Wrong password for username %s", user.Username))
+	// redirect if invalid
+	if user.Hd != "uwblueprint.org" && user.Hd != "theshoeproject.online" {
+		http.Redirect(w, r, "/api/unauthorized", http.StatusTemporaryRedirect)
 	}
 
-	jwtExpiry, err := config.GetTokenExpiryDuration()
+	// if valid, save to db and create jwt token
+	api.database.FirstOrCreate(&user)
+	jwtToken, err := generateJWTToken(user.Email)
 	if err != nil {
 		return rest.ErrInternal(api.logger, err)
 	}
 
+	return rest.JSONStatusOK(jwtToken)
```

Comment:
I thought we were going to put the JWT token in the cookie of the response?

Reply:
Also we should redirect users to the all stories page if the authentication is successful. Since this page currently doesn't exist we can redirect them to `/` (i.e homepage) and create another (separate) pr to change this once all stories has been implemented.

Reply:
ok, this will be done in next pr! didn't want to touch jwt flow in the same pr...